File: AssetManager/Data/DataFunctions/DatabaseHelperFunctions/FTPFunctions.vb
Public Class FtpFunctions

#Region "Fields"

    Private FTPComms As New FtpComms

#End Region

#Region "Methods"

    Public Function DeleteFtpAttachment(fileUID As String, fKey As String) As Boolean
        Dim resp As Net.FtpWebResponse = Nothing
        Try
            resp = DirectCast(FTPComms.ReturnFtpResponse("ftp://" & ServerInfo.MySQLServerIP & "/attachments/" & ServerInfo.CurrentDataBase.ToString & "/" & fKey & "/" & fileUID, Net.WebRequestMethods.Ftp.DeleteFile), Net.FtpWebResponse)
            If resp.StatusCode = Net.FtpStatusCode.FileActionOK Then
                Return True
            Else
                Return False
            End If
        Catch ex As Exception
            Return ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
        End Try
    End Function

    Public Function DeleteFtpFolder(folderUID As String) As Boolean
        Try
            Dim files = ListDirectory("ftp://" & ServerInfo.MySQLServerIP & "/attachments/" & ServerInfo.CurrentDataBase.ToString & "/" & folderUID & "/")
            Dim i As Integer = 0
            For Each file As String In files   'delete each file counting for successes
                If DeleteFtpAttachment(file, folderUID) Then i += 1
            Next
            If files.Count = i Then ' if successful deletetions = total # of files, delete the directory
                Using deleteResp = DirectCast(FTPComms.ReturnFtpResponse("ftp://" & ServerInfo.MySQLServerIP & "/attachments/" & ServerInfo.CurrentDataBase.ToString & "/" & folderUID, Net.WebRequestMethods.Ftp.RemoveDirectory), Net.FtpWebResponse)
                    If deleteResp.StatusCode = Net.FtpStatusCode.FileActionOK Then
                        Return True
                    End If
                End Using
            End If
            Return False
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
            Return False
        End Try
    End Function

    Public Function HasFtpFolder(itemGUID As String) As Boolean
        Try
            Using resp As Net.FtpWebResponse = DirectCast(FTPComms.ReturnFtpResponse("ftp://" & ServerInfo.MySQLServerIP & "/attachments/" & ServerInfo.CurrentDataBase.ToString & "/" & itemGUID & "/", Net.WebRequestMethods.Ftp.ListDirectory), Net.FtpWebResponse)
                If resp.StatusCode = Net.FtpStatusCode.OpeningData Then Return True
            End Using
            Return False
        Catch ex As Exception
            Return False
        End Try
    End Function

    Public Sub ScanAttachements()
        Try
            Logger("***********************************")
            Logger("******Attachment Scan Results******")

            Dim FTPDirs = ListDirectory("ftp://" & ServerInfo.MySQLServerIP & "/attachments/" & ServerInfo.CurrentDataBase.ToString & "/")
            Dim FTPFiles = ListFTPFiles(FTPDirs)
            Dim SQLFiles = ListSQLFiles()

            Dim MissingFTPDirs = ListMissingFTPDirs(FTPDirs)
            Dim MissingSQLDirs = ListMissingSQLDirs(SQLFiles, FTPDirs)

            Dim MissingFTPFiles = ListMissingFTPFiles(SQLFiles, FTPFiles)
            Dim MissingSQLFiles = ListMissingSQLFiles(SQLFiles, FTPFiles)

            If MissingFTPDirs.Count > 0 Or MissingSQLDirs.Count > 0 Or MissingFTPFiles.Count > 0 Or MissingSQLFiles.Count > 0 Then
                Dim StatsText As String = ""
                Logger("Orphan Files/Directories found!")
                StatsText = "Orphan Files/Directories found!  Do you want to delete the corrupt SQL/FTP entries?

FTP: 
Missing Dirs: " & MissingFTPDirs.Count & "
Missing Files: " & MissingFTPFiles.Count & "

SQL:
Missing Dirs: " & MissingSQLDirs.Count & "
Missing Files: " & MissingSQLFiles.Count

                Dim blah = Message(StatsText, vbYesNo + vbExclamation, "Orphans Found")
                If blah = MsgBoxResult.Yes Then
                    'clean it up
                    Logger("Cleaning attachments...")
                    Dim itemsCleaned As Integer = 0
                    itemsCleaned += CleanFTPFiles(MissingFTPFiles)
                    itemsCleaned += CleanFTPDirs(MissingFTPDirs)
                    itemsCleaned += CleanSQLFiles(MissingSQLFiles)
                    itemsCleaned += CleanSQLEntries(MissingSQLDirs)
                    Message("Cleaned " & itemsCleaned & " orphans.")
                    ScanAttachements()
                End If
            Else
                Logger("No Orphans Found.")
                Message("No issues found.", vbOKOnly + vbInformation, "Scan OK")
            End If
            Logger("**********End Scan Results*********")
            Logger("***********************************")
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
        End Try
    End Sub

    ''' <summary>
    ''' Checks if supplied UID exists in Devices or Sibi tables.
    ''' </summary>
    ''' <returns></returns>
    Private Function CheckForPrimaryItem(itemUID As String) As Boolean
        Dim exists As Boolean = False
        If AssetFunc.GetSqlValue(DevicesCols.TableName, DevicesCols.DeviceUID, itemUID, DevicesCols.DeviceUID) <> "" Then exists = True
        If AssetFunc.GetSqlValue(SibiRequestCols.TableName, SibiRequestCols.UID, itemUID, SibiRequestCols.UID) <> "" Then exists = True
        Return exists
    End Function

    Private Function CleanSQLFiles(missingSQLFiles As List(Of AttachScanInfo)) As Integer
        If missingSQLFiles.Count > 0 Then
            Dim DeviceTable As New DeviceAttachmentsCols
            Dim SibiTable As New SibiAttachmentsCols
            Dim deletions As Integer = 0
            For Each sqlItem In missingSQLFiles
                Dim DeviceRows = DBFunc.GetDatabase.ExecuteQuery("DELETE FROM " & DeviceTable.TableName & " WHERE " & DeviceTable.FileUID & "='" & sqlItem.FileUID & "'")
                If DeviceRows > 0 Then
                    deletions += DeviceRows
                    Logger("Deleted Device SQL File: " & sqlItem.FKey & "/" & sqlItem.FileUID)
                End If

                Dim SibiRows = DBFunc.GetDatabase.ExecuteQuery("DELETE FROM " & SibiTable.TableName & " WHERE " & SibiTable.FileUID & "='" & sqlItem.FileUID & "'")
                If SibiRows > 0 Then
                    deletions += SibiRows
                    Logger("Deleted Sibi SQL File: " & sqlItem.FKey & "/" & sqlItem.FileUID)
                End If
            Next
            Return deletions
        End If
        Return 0
    End Function

    Private Function CleanSQLEntries(missingSQLDirs As List(Of AttachScanInfo)) As Integer
        If missingSQLDirs.Count > 0 Then
            Dim DeviceTable As New DeviceAttachmentsCols
            Dim SibiTable As New SibiAttachmentsCols
            Dim deletions As Integer = 0
            For Each sqlItem In missingSQLDirs
                If Not CheckForPrimaryItem(sqlItem.FKey) Then

                    Dim DeviceRows = DBFunc.GetDatabase.ExecuteQuery("DELETE FROM " & DeviceTable.TableName & " WHERE " & DeviceTable.FKey & "='" & sqlItem.FKey & "'")
                    If DeviceRows > 0 Then
                        deletions += DeviceRows
                        Logger("Deleted " & DeviceRows & " Device SQL Entries For: " & sqlItem.FKey)
                    End If

                    Dim SibiRows = DBFunc.GetDatabase.ExecuteQuery("DELETE FROM " & SibiTable.TableName & " WHERE " & SibiTable.FKey & "='" & sqlItem.FKey & "'")
                    If SibiRows > 0 Then
                        deletions += SibiRows
                        Logger("Deleted " & SibiRows & " Sibi SQL Entries For: " & sqlItem.FKey)
                    End If

                End If
            Next
            Return deletions
        End If
        Return 0
    End Function

    Private Function CleanFTPFiles(missingFTPFiles As List(Of AttachScanInfo)) As Integer
        Dim deletions As Integer = 0
        For Each file In missingFTPFiles
            If DeleteFtpAttachment(file.FileUID, file.FKey) Then
                deletions += 1
                Logger("Deleted SQL File: " & file.FKey & "/" & file.FileUID)
            End If
        Next
        Return deletions
    End Function

    Private Function CleanFTPDirs(missingFTPDirs As List(Of String)) As Integer
        Dim deletions As Integer = 0
        For Each fDir In missingFTPDirs
            If DeleteFtpFolder(fDir) Then
                deletions += 1
                Logger("Deleted SQL Directory: " & fDir)
            End If
        Next
        Return deletions
    End Function

    ''' <summary>
    ''' Returns list of SQL entries not found in FTP directory list.
    ''' </summary>
    ''' <param name="sqlFiles"></param>
    ''' <param name="ftpDirs"></param>
    ''' <returns></returns>
    Private Function ListMissingSQLDirs(sqlFiles As List(Of AttachScanInfo), ftpDirs As List(Of String)) As List(Of AttachScanInfo)
        Dim MissingDirs As New List(Of AttachScanInfo)
        For Each SQLfile In sqlFiles
            Dim match As Boolean = False
            ftpDirs.ForEach(Sub(f) If SQLfile.FKey = f Then match = True)
            If Not match Then
                If Not MissingDirs.Exists(Function(d) SQLfile.FKey = d.FKey) Then MissingDirs.Add(SQLfile)
            End If
        Next
        MissingDirs.ForEach(Sub(d) Logger("Orphan SQL Dir Found: " & d.FKey))
        Return MissingDirs
    End Function

    ''' <summary>
    ''' Returns list of FTP dirs that do not have an associated Device or Sibi request.
    ''' </summary>
    ''' <param name="ftpDirs"></param>
    ''' <returns></returns>
    Private Function ListMissingFTPDirs(ftpDirs As List(Of String)) As List(Of String)
        Dim MissingDirs = ftpDirs.FindAll(Function(f) Not CheckForPrimaryItem(f))
        MissingDirs.ForEach(Sub(f) Logger("Orphan FTP Dir Found: " & f))
        Return MissingDirs
    End Function

    ''' <summary>
    ''' Returns list of SQL files not found in FTP file list.
    ''' </summary>
    ''' <param name="sqlFiles"></param>
    ''' <param name="ftpFiles"></param>
    ''' <returns></returns>
    Private Function ListMissingSQLFiles(sqlFiles As List(Of AttachScanInfo), ftpFiles As List(Of AttachScanInfo)) As List(Of AttachScanInfo)
        Dim MissingFiles = sqlFiles.Except(ftpFiles).ToList
        MissingFiles.ForEach(Sub(f) Logger("Orphan SQL File Found: " & f.FKey & "/" & f.FileUID))
        Return sqlFiles.Except(ftpFiles).ToList
    End Function

    ''' <summary>
    ''' Returns list of FTP files not found in SQL file list.
    ''' </summary>
    ''' <param name="sqlFiles"></param>
    ''' <param name="ftpFiles"></param>
    ''' <returns></returns>
    Private Function ListMissingFTPFiles(sqlFiles As List(Of AttachScanInfo), ftpFiles As List(Of AttachScanInfo)) As List(Of AttachScanInfo)
        Dim MissingFiles = ftpFiles.Except(sqlFiles).ToList
        MissingFiles.ForEach(Sub(f) Logger("Orphan FTP File Found: " & f.FKey & "/" & f.FileUID))
        Return MissingFiles
    End Function

    Private Function ListFTPFiles(ftpDirs As List(Of String)) As List(Of AttachScanInfo)
        Dim FTPFileList As New List(Of AttachScanInfo)
        For Each fDir In ftpDirs
            For Each file In ListDirectory("ftp://" & ServerInfo.MySQLServerIP & "/attachments/" & ServerInfo.CurrentDataBase.ToString & "/" & fDir & "/")
                FTPFileList.Add(New AttachScanInfo(fDir, file))
            Next
        Next
        Return FTPFileList
    End Function

    Private Function ListSQLFiles() As List(Of AttachScanInfo)
        Dim DeviceTable As New DeviceAttachmentsCols
        Dim SibiTable As New SibiAttachmentsCols
        Dim SQLFileList As New List(Of AttachScanInfo)

        Dim devFiles = DBFunc.GetDatabase.DataTableFromQueryString("SELECT * FROM " & DeviceTable.TableName)
        For Each file As DataRow In devFiles.Rows
            SQLFileList.Add(New AttachScanInfo(file.Item(DeviceTable.FKey).ToString, file.Item(DeviceTable.FileUID).ToString))
        Next

        Dim sibiFiles = DBFunc.GetDatabase.DataTableFromQueryString("SELECT * FROM " & SibiTable.TableName)
        For Each file As DataRow In sibiFiles.Rows
            SQLFileList.Add(New AttachScanInfo(file.Item(SibiTable.FKey).ToString, file.Item(SibiTable.FileUID).ToString))
        Next

        Return SQLFileList
    End Function

    Private Function ListDirectory(uri As String) As List(Of String)
        Try
            Using resp = DirectCast(FTPComms.ReturnFtpResponse(uri, Net.WebRequestMethods.Ftp.ListDirectory), Net.FtpWebResponse)
                Dim responseStream As System.IO.Stream = resp.GetResponseStream
                Using reader As IO.StreamReader = New IO.StreamReader(responseStream)
                    Dim files As New List(Of String)
                    While Not reader.EndOfStream 'collect list of files in directory
                        files.Add(reader.ReadLine)
                    End While
                    Return files
                End Using
            End Using
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
            Return Nothing
        End Try
    End Function

    Private Structure AttachScanInfo
        Public FKey As String
        Public FileUID As String
        Sub New(FKey As String, FileUID As String)
            Me.FKey = FKey
            Me.FileUID = FileUID
        End Sub
    End Structure
#End Region

End Class

File: AssetManager/Data/DataFunctions/DatabaseHelperFunctions/FTPFunctions.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using System.Net;
using System.Windows.Forms;

namespace AssetManager
{
    public class FtpFunctions
    {
        #region Fields

        private FtpComms FTPComms = new FtpComms();

        #endregion Fields

        #region Methods

        public bool DeleteFtpAttachment(string fileUID, string fKey)
        {
            FtpWebResponse resp = null;
            try
            {
                resp = (FtpWebResponse)(FTPComms.ReturnFtpResponse("ftp://" + ServerInfo.MySQLServerIP + "/attachments/" + ServerInfo.CurrentDataBase.ToString() + "/" + fKey + "/" + fileUID, WebRequestMethods.Ftp.DeleteFile));
                if (resp.StatusCode == FtpStatusCode.FileActionOK)
                {
                    return true;
                }
                else
                {
                    return false;
                }
            }
            catch (Exception ex)
            {
                return ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
            }
        }

        public bool DeleteFtpFolder(string folderUID)
        {
            try
            {
                var files = ListDirectory("ftp://" + ServerInfo.MySQLServerIP + "/attachments/" + ServerInfo.CurrentDataBase.ToString() + "/" + folderUID + "/");
                int i = 0;
                foreach (string file in files) //delete each file counting for successes
                {
                    if (DeleteFtpAttachment(file, folderUID))
                    {
                        i++;
                    }
                }
                if (files.Count == i) // if successful deletions = total # of files, delete the directory
                {
                    using (var deleteResp = (FtpWebResponse)(FTPComms.ReturnFtpResponse("ftp://" + ServerInfo.MySQLServerIP + "/attachments/" + ServerInfo.CurrentDataBase.ToString() + "/" + folderUID, WebRequestMethods.Ftp.RemoveDirectory)))
                    {
                        if (deleteResp.StatusCode == FtpStatusCode.FileActionOK)
                        {
                            return true;
                        }
                    }
                }
                return false;
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
                return false;
            }
        }

        public bool HasFtpFolder(string itemGUID)
        {
            try
            {
                using (FtpWebResponse resp = (FtpWebResponse)(FTPComms.ReturnFtpResponse("ftp://" + ServerInfo.MySQLServerIP + "/attachments/" + ServerInfo.CurrentDataBase.ToString() + "/" + itemGUID + "/", WebRequestMethods.Ftp.ListDirectory)))
                {
                    if (resp.StatusCode == FtpStatusCode.OpeningData)
                    {
                        return true;
                    }
                }

                return false;
            }
            catch (Exception)
            {
                return false;
            }
        }

        public void ScanAttachements()
        {
            try
            {
                Logging.Logger("***********************************");
                Logging.Logger("******Attachment Scan Results******");

                var FTPDirs = ListDirectory("ftp://" + ServerInfo.MySQLServerIP + "/attachments/" + ServerInfo.CurrentDataBase.ToString() + "/");
                var FTPFiles = ListFTPFiles(FTPDirs);
                var SQLFiles = ListSQLFiles();

                var MissingFTPDirs = ListMissingFTPDirs(FTPDirs);
                var MissingSQLDirs = ListMissingSQLDirs(SQLFiles, FTPDirs);

                var MissingFTPFiles = ListMissingFTPFiles(SQLFiles, FTPFiles);
                var MissingSQLFiles = ListMissingSQLFiles(SQLFiles, FTPFiles);

                if (MissingFTPDirs.Count > 0 || MissingSQLDirs.Count > 0 || MissingFTPFiles.Count > 0 || MissingSQLFiles.Count > 0)
                {
                    string StatsText = "";
                    Logging.Logger("Orphan Files/Directories found!");
                    StatsText = @"Orphan Files/Directories found!  Do you want to delete the corrupt SQL/FTP entries?

FTP:
Missing Dirs: " + MissingFTPDirs.Count + @"
Missing Files: " + MissingFTPFiles.Count + @"

SQL:
Missing Dirs: " + MissingSQLDirs.Count + @"
Missing Files: " + MissingSQLFiles.Count;

                    var blah = OtherFunctions.Message(StatsText, (int)MessageBoxButtons.YesNo + (int)MessageBoxIcon.Exclamation, "Orphans Found");
                    if (blah == DialogResult.Yes)
                    {
                        //clean it up
                        Logging.Logger("Cleaning attachments...");
                        int itemsCleaned = 0;
                        itemsCleaned += CleanFTPFiles(MissingFTPFiles);
                        itemsCleaned += CleanFTPDirs(MissingFTPDirs);
                        itemsCleaned += CleanSQLFiles(MissingSQLFiles);
                        itemsCleaned += CleanSQLEntries(MissingSQLDirs);
                        OtherFunctions.Message("Cleaned " + itemsCleaned.ToString() + " orphans.");
                        ScanAttachements();
                    }
                }
                else
                {
                    Logging.Logger("No Orphans Found.");
                    OtherFunctions.Message("No issues found.", (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Information, "Scan OK");
                }
                Logging.Logger("**********End Scan Results*********");
                Logging.Logger("***********************************");
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
            }
        }

        /// <summary>
        /// Checks if supplied UID exists in Devices or Sibi tables.
        /// </summary>
        /// <returns></returns>
        private bool CheckForPrimaryItem(string itemUID)
        {
            bool exists = false;
            if (GlobalInstances.AssetFunc.GetSqlValue(DevicesCols.TableName, DevicesCols.DeviceUID, itemUID, DevicesCols.DeviceUID) != "")
            {
                exists = true;
            }
            if (GlobalInstances.AssetFunc.GetSqlValue(SibiRequestCols.TableName, SibiRequestCols.UID, itemUID, SibiRequestCols.UID) != "")
            {
                exists = true;
            }
            return exists;
        }

        private int CleanSQLFiles(List<AttachScanInfo> missingSQLFiles)
        {
            if (missingSQLFiles.Count > 0)
            {
                DeviceAttachmentsCols DeviceTable = new DeviceAttachmentsCols();
                SibiAttachmentsCols SibiTable = new SibiAttachmentsCols();
                int deletions = 0;
                foreach (var sqlItem in missingSQLFiles)
                {
                    var DeviceRows = DBFactory.GetDatabase().ExecuteQuery("DELETE FROM " + DeviceTable.TableName + " WHERE " + DeviceTable.FileUID + "='" + sqlItem.FileUID + "'");
                    if (DeviceRows > 0)
                    {
                        deletions += DeviceRows;
                        Logging.Logger("Deleted Device SQL File: " + sqlItem.FKey + "/" + sqlItem.FileUID);
                    }

                    var SibiRows = DBFactory.GetDatabase().ExecuteQuery("DELETE FROM " + SibiTable.TableName + " WHERE " + SibiTable.FileUID + "='" + sqlItem.FileUID + "'");
                    if (SibiRows > 0)
                    {
                        deletions += SibiRows;
                        Logging.Logger("Deleted Sibi SQL File: " + sqlItem.FKey + "/" + sqlItem.FileUID);
                    }
                }
                return deletions;
            }
            return 0;
        }

        private int CleanSQLEntries(List<AttachScanInfo> missingSQLDirs)
        {
            if (missingSQLDirs.Count > 0)
            {
                DeviceAttachmentsCols DeviceTable = new DeviceAttachmentsCols();
                SibiAttachmentsCols SibiTable = new SibiAttachmentsCols();
                int deletions = 0;
                foreach (var sqlItem in missingSQLDirs)
                {
                    if (!CheckForPrimaryItem(sqlItem.FKey))
                    {
                        var DeviceRows = DBFactory.GetDatabase().ExecuteQuery("DELETE FROM " + DeviceTable.TableName + " WHERE " + DeviceTable.FKey + "='" + sqlItem.FKey + "'");
                        if (DeviceRows > 0)
                        {
                            deletions += DeviceRows;
                            Logging.Logger("Deleted " + DeviceRows.ToString() + " Device SQL Entries For: " + sqlItem.FKey);
                        }

                        var SibiRows = DBFactory.GetDatabase().ExecuteQuery("DELETE FROM " + SibiTable.TableName + " WHERE " + SibiTable.FKey + "='" + sqlItem.FKey + "'");
                        if (SibiRows > 0)
                        {
                            deletions += SibiRows;
                            Logging.Logger("Deleted " + SibiRows.ToString() + " Sibi SQL Entries For: " + sqlItem.FKey);
                        }
                    }
                }
                return deletions;
            }
            return 0;
        }

        private int CleanFTPFiles(List<AttachScanInfo> missingFTPFiles)
        {
            int deletions = 0;
            foreach (var file in missingFTPFiles)
            {
                if (DeleteFtpAttachment(file.FileUID, file.FKey))
                {
                    deletions++;
                    Logging.Logger("Deleted SQL File: " + file.FKey + "/" + file.FileUID);
                }
            }
            return deletions;
        }

        private int CleanFTPDirs(List<string> missingFTPDirs)
        {
            int deletions = 0;
            foreach (var fDir in missingFTPDirs)
            {
                if (DeleteFtpFolder(fDir))
                {
                    deletions++;
                    Logging.Logger("Deleted SQL Directory: " + fDir);
                }
            }
            return deletions;
        }

        /// <summary>
        /// Returns list of SQL entries not found in FTP directory list.
        /// </summary>
        /// <param name="sqlFiles"></param>
        /// <param name="ftpDirs"></param>
        /// <returns></returns>
        private List<AttachScanInfo> ListMissingSQLDirs(List<AttachScanInfo> sqlFiles, List<string> ftpDirs)
        {
            List<AttachScanInfo> MissingDirs = new List<AttachScanInfo>();
            foreach (var SQLfile in sqlFiles)
            {
                bool match = false;
                ftpDirs.ForEach(f =>
                {
                    if (SQLfile.FKey == f)
                    {
                        match = true;
                    }
                });
                if (!match)
                {
                    if (!MissingDirs.Exists(d => SQLfile.FKey == d.FKey))
                    {
                        MissingDirs.Add(SQLfile);
                    }
                }
            }
            MissingDirs.ForEach(d => Logging.Logger("Orphan SQL Dir Found: " + d.FKey));
            return MissingDirs;
        }

        /// <summary>
        /// Returns list of FTP dirs that do not have an associated Device or Sibi request.
        /// </summary>
        /// <param name="ftpDirs"></param>
        /// <returns></returns>
        private List<string> ListMissingFTPDirs(List<string> ftpDirs)
        {
            var MissingDirs = ftpDirs.FindAll(f => !CheckForPrimaryItem(f));
            MissingDirs.ForEach(f => Logging.Logger("Orphan FTP Dir Found: " + f));
            return MissingDirs;
        }

        /// <summary>
        /// Returns list of SQL files not found in FTP file list.
        /// </summary>
        /// <param name="sqlFiles"></param>
        /// <param name="ftpFiles"></param>
        /// <returns></returns>
        private List<AttachScanInfo> ListMissingSQLFiles(List<AttachScanInfo> sqlFiles, List<AttachScanInfo> ftpFiles)
        {
            var MissingFiles = sqlFiles.Except(ftpFiles).ToList();
            MissingFiles.ForEach(f => Logging.Logger("Orphan SQL File Found: " + f.FKey + "/" + f.FileUID));
            return sqlFiles.Except(ftpFiles).ToList();
        }

        /// <summary>
        /// Returns list of FTP files not found in SQL file list.
        /// </summary>
        /// <param name="sqlFiles"></param>
        /// <param name="ftpFiles"></param>
        /// <returns></returns>
        private List<AttachScanInfo> ListMissingFTPFiles(List<AttachScanInfo> sqlFiles, List<AttachScanInfo> ftpFiles)
        {
            var MissingFiles = ftpFiles.Except(sqlFiles).ToList();
            MissingFiles.ForEach(f => Logging.Logger("Orphan FTP File Found: " + f.FKey + "/" + f.FileUID));
            return MissingFiles;
        }

        private List<AttachScanInfo> ListFTPFiles(List<string> ftpDirs)
        {
            List<AttachScanInfo> FTPFileList = new List<AttachScanInfo>();
            foreach (var fDir in ftpDirs)
            {
                foreach (var file in ListDirectory("ftp://" + ServerInfo.MySQLServerIP + "/attachments/" + ServerInfo.CurrentDataBase.ToString() + "/" + fDir + "/"))
                {
                    FTPFileList.Add(new AttachScanInfo(fDir, file));
                }
            }
            return FTPFileList;
        }

        private List<AttachScanInfo> ListSQLFiles()
        {
            DeviceAttachmentsCols DeviceTable = new DeviceAttachmentsCols();
            SibiAttachmentsCols SibiTable = new SibiAttachmentsCols();
            List<AttachScanInfo> SQLFileList = new List<AttachScanInfo>();

            var devFiles = DBFactory.GetDatabase().DataTableFromQueryString("SELECT * FROM " + DeviceTable.TableName);
            foreach (DataRow file in devFiles.Rows)
            {
                SQLFileList.Add(new AttachScanInfo(file[DeviceTable.FKey].ToString(), file[DeviceTable.FileUID].ToString()));
            }

            var sibiFiles = DBFactory.GetDatabase().DataTableFromQueryString("SELECT * FROM " + SibiTable.TableName);
            foreach (DataRow file in sibiFiles.Rows)
            {
                SQLFileList.Add(new AttachScanInfo(file[SibiTable.FKey].ToString(), file[SibiTable.FileUID].ToString()));
            }

            return SQLFileList;
        }

        private List<string> ListDirectory(string uri)
        {
            try
            {
                using (var resp = (FtpWebResponse)(FTPComms.ReturnFtpResponse(uri, WebRequestMethods.Ftp.ListDirectory)))
                {
                    System.IO.Stream responseStream = resp.GetResponseStream();
                    using (StreamReader reader = new StreamReader(responseStream))
                    {
                        List<string> files = new List<string>();
                        while (!reader.EndOfStream) //collect list of files in directory
                        {
                            files.Add(reader.ReadLine());
                        }
                        return files;
                    }
                }
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
                return default(List<string>);
            }
        }

        private struct AttachScanInfo
        {
            public string FKey;
            public string FileUID;

            public AttachScanInfo(string FKey, string FileUID)
            {
                this.FKey = FKey;
                this.FileUID = FileUID;
            }
        }

        #endregion Methods
    }
}